### PR TITLE
ALF will save a visualization of the algorithm module structure

### PR DESCRIPTION
**Pull request by breakds**
# Motivation

For more complex algorithms, it is useful to be able to visualize the hierarchy of the algorithm, sub algorithms and network structures so that we can make sure the components are correctly wired.

# Solution

Save a visualized graph of algorithm hierarchy when summarizing the training settings.

This is the first implementation and many can be improved from here.

# Testing

An example generated graph for PPG + Transformer + MetaDrive:

algorithm_sturcture

### CODE REVIEW

File: alf/trainers/policy_trainer.py [hunk 1]
```diff
@@ -246,6 +310,11 @@ def _markdownify(paragraph):
             alf.summary.text('diff', _markdownify(git_utils.get_diff()))
             alf.summary.text('seed', str(self._random_seed))
 
+            # Save a rendered directed graph of the algorithm to the root
+            # directory.
+            _visualize_alf_tree(self._algorithm).render(
+                Path(flags.FLAGS.root_dir, 'algorithm_sturcture'))
```
Comment (hnyu) [inline alf/trainers/policy_trainer.py]: should probably use `self._root_dir`

File: alf/trainers/policy_trainer.py [hunk 2]
```diff
@@ -246,6 +310,11 @@ def _markdownify(paragraph):
             alf.summary.text('diff', _markdownify(git_utils.get_diff()))
             alf.summary.text('seed', str(self._random_seed))
 
+            # Save a rendered directed graph of the algorithm to the root
+            # directory.
+            _visualize_alf_tree(self._algorithm).render(
```
Comment (hnyu) [inline alf/trainers/policy_trainer.py]: Do you think it's better to put this image in Tensorboard instead?

Comment (breakds) [reply alf/trainers/policy_trainer.py]: Should be able to do that on top of saving it to the `root_dir`. I could not find a way to directly render graphviz to tensor, so the procedure has to be saving it to the disk, load the image as tensor and summarize it. I'll give it a try tomorrow.

**Comment by breakds** [comment]
Apparently also need to install `graphviz` :) Will make that happen in another PR.

**Comment by hnyu** [comment]
> Apparently also need to install `graphviz` :) Will make that happen in another PR.

I think in order to insteall graphviz, you need to modify the docker file to use apt install.

**Comment by breakds** [comment]
Updated the PR with

1. Also send the graph to tensorboard.
2. If `graphviz` is not present, warn instead of panic.

tensorboard_alg_structure

File: alf/examples/metadrive/ppg_metadrive_conf.py [hunk 3]
```diff
@@ -30,7 +30,7 @@
 
 # Environment Configuration
 alf.config(
-    'create_environment', env_name='BirdEye', num_parallel_environments=36)
+    'create_environment', env_name='BirdEye', num_parallel_environments=3)
```
Comment (hnyu) [inline alf/examples/metadrive/ppg_metadrive_conf.py]: Why having this change to the conf.py?

Comment (breakds) [reply alf/examples/metadrive/ppg_metadrive_conf.py]: Fixed. This is included by accident.

**Review by hnyu — APPROVED** [review]

**Pull request merged**